plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Stablecoin.hs

```diff
@@ -0,0 +1,385 @@
+
+{-# INLINEABLE equity #-}
+-- | The bank's equity (what's left of the reserves after subtracting
+--   liabilities).
+equity ::
+    BankState
+    -> ConversionRate
+    -> BC (Ratio Integer)
+equity r@BankState{bsReserves=BC reserves} cr =
+    let BC l = liabilities r cr
+    in BC (fromInteger reserves - l)
+
+-- | Stablecoin parameters.
+data Stablecoin =
+    Stablecoin
+        { scOracle                  :: PubKey -- ^ Public key of the oracle that provides exchange rates
+        , scFee                     :: Ratio Integer -- ^ Fee charged by bank for transactions. Calculated as a fraction of the total transaction volume in base currency.
+        , scMinReserveRatio         :: Ratio Integer -- ^ The minimum ratio of reserves to liabilities
+        , scReservecoinDefaultPrice :: BC Integer -- ^ The price of a single reservecoin if no reservecoins have been issued
+        , scBaseCurrency            :: (CurrencySymbol, TokenName) -- ^ The base currency. Value of this currency will be locked by the stablecoin state machine instance
+        , scStablecoinTokenName     :: TokenName -- ^ 'TokenName' of the stablecoin
+        , scReservecoinTokenName    :: TokenName -- ^ 'TokenName' of the reservecoin
+        }
+    deriving stock (Generic, Haskell.Eq, Haskell.Show)
+    deriving anyclass (ToJSON, FromJSON)
+
+{-# INLINEABLE minReserve #-}
+-- | Minimum number of base currency coins held by the bank
+minReserve :: Stablecoin -> ConversionRate -> BankState -> BC (Ratio Integer)
+minReserve Stablecoin{scMinReserveRatio} cr BankState{bsStablecoins=SC sc} =
+    let BC r = convert cr (PC $ fromInteger sc)
+    in BC $ scMinReserveRatio * r
+
+{-# INLINEABLE reservecoinNominalPrice #-}
+-- | Price of a single reservecoin in base currency
+reservecoinNominalPrice :: Stablecoin -> BankState -> ConversionRate -> BC (Ratio Integer)
+reservecoinNominalPrice Stablecoin{scReservecoinDefaultPrice} bankState@BankState{bsReservecoins=RC rc} cr =
+    if rc /= 0
+        then
+            let BC e = equity bankState cr
+            in BC (e * R.recip (fromInteger rc))
+        else fmap fromInteger scReservecoinDefaultPrice
+
+{-# INLINEABLE stablecoinNominalPrice #-}
+-- | Price of a single stablecoin in base currency. If the banks' liabilities
+--   exceed its reserves then 'stablecoinNominalPrice' is zero.
+stablecoinNominalPrice :: BankState -> ConversionRate -> BC (Ratio Integer)
+stablecoinNominalPrice bankState@BankState{bsStablecoins=SC sc} cr =
+    let BC p = convert cr (PC $ fromInteger 1)
+        BC l = liabilities bankState cr
+    in if sc == zero
+        then BC p
+        else BC $ min p l
+
+-- | Action that can be performed on the stablecoin contract.
+data SCAction
+    = MintStablecoin (SC Integer) -- ^ Create a number stablecoins, depositing the matching amount of base currency
+    | RedeemStablecoin (SC Integer) -- ^ Destroy a number of stablecoins, releasing the matching amount of base currency
```

Comment:
Could merge these with a single argument that can be negative! Otherwise I think we should have some checks that this isn't negative...

Reply:
I actually think this would all work fine with generalized `*NominalPrice` functions: they could tell you the total price to change the supply from `current` to `current + delta`, where delta could be negative, and then the price would be negative, and it might all be quite nice.

Reply:
Deleting `RedeemStablecoin` and `RedeemReservecoin` works without any changes actually, because we use `baseCurrencyValue` to compute the new value after the state change, so we don't need to convert the delta to a `Value`